### PR TITLE
⚡️ Improve logging time

### LINKED ISSUE #41: See if login time can be reduced
Login time can be quite long and this is especially annoying for this app because:
- time in the app is very short, so login time is significant
- most user will come no more than once a month which means most logins will be cold (no token renewal possible)

As far as I could see most of the time is waiting for the HTTP request to Auth0, so I'm not sure how we can improve, but I'd be happy for someone to have another look.

As an alternative, I guess we could take the optimistic UI approach and show the vote screen immediately after the user comes back from Auth0, not waiting for the HTTP response, and then only if login is rejected the app can show an alert and reload, or something like that.

### PR DESCRIPTION
**Pull request by jsulpis**
Close #41 

I made quite a few changes, I hope it is not too much for one PR. But the logging time (at least  perceived) is much shorter. 

Here is what I did:

- remove the call to the auth0 API `authenticationClient.getProfile()` in the `exchangeToken` function, which was not necessary since all we got from it was the user email, which we already have from the auth0 authentication. The call to this function was one of the longest so I think we can save several seconds on the production app just by removing it.
- change the order of some operations: use optimistic UI to display the main content right after the auth0 authentication, and display the rest after the Firebase authentication. Also fetch the list of agencies at the very end since it is in the second page so we don't need it straight away.
- display a little loader during the Firebase authentication when the main content is already there, to let the user know that there is still something happening in the background so they are not frustrated if an error message shows up (like campaign closed)

### CODE REVIEW

**Comment by hgwood** [comment]
Wow! Sounds awesome. I'll try to review and test it this week.

File: ui/src/index.ts [hunk 1]
```diff
@@ -71,9 +71,11 @@ window.addEventListener("load", async function () {
           strokeMiterlimit="10"
         />
       </svg>
-    </div>
-    Hold on, we're retrieving the data you requested ...
-  </h2>`;
+    </div>`;
+
+  const statsLoader = `<h2>${htmlLoader} Hold on, we're retrieving the data you requested ...</h2>`;
```
Comment (EmrysMyrddin) [inline ui/src/index.ts:76]: For accessibility, can we use something else than a h2 to display information that is not really a title ?

**Review by EmrysMyrddin — APPROVED** [review]
LGTM

**Comment by hgwood** [comment]
I'm getting an error when running the app locally, I'm investigating.

image

**Review by hgwood — CHANGES_REQUESTED** [review]
I'm afraid removing the call to Auth0 is not that easy. Your proposal is not secured. Anyone can call `exchangeToken` with any values for `userId`, `userEmail`, `accessToken` and get authenticated to Firebase. In the `master` version, the value for `accessToken` is validated against Auth0, which ensures that the token is legitimate and that the user really did authenticate to Auth0. This is why Auth0 is called, not just to get the email address. 🙂

File: ui/src/index.ts (hunk 1 above)
Comment (jsulpis) [reply ui/src/index.ts]: I put a div instead

**Comment by jsulpis** [comment]
> I'm afraid removing the call to Auth0 is not that easy. Your proposal is not secured. Anyone can call `exchangeToken` with any values for `userId`, `userEmail`, `accessToken` and get authenticated to Firebase. In the `master` version, the value for `accessToken` is validated against Auth0, which ensures that the token is legitimate and that the user really did authenticate to Auth0. This is why Auth0 is called, not just to get the email address. 🙂

Indeed I forgot it was in an API function and thus not redundant with the authentication. I dropped the commit.

Hopefully with the optimistic UI it should still feel faster, but the total logging time will remain roughly the same (very long). So I am not sure if this PR can still close the associated issue ? It strange that some API calls take much more time on the production app than locally, but I won't be able to further analyse this issue.

**Comment by hgwood** [comment]
I'll come back to this PR next week :)

**Comment by hgwood** [comment]
I might have solution for the Auth0 call, see https://github.com/Zenika/humeur-du-mois-2018/issues/41#issuecomment-710131946, but let's keep that for another PR.

**Comment by hgwood** [comment]
Finally got around to test this. It works nicely, thank you. I think it can be improved further. In the current form, the Stats button in the navbar is unresponsive while authenticating to Firebase. I don't think this has to happen. I'm gonna to test a few adjustment to see what I can do.

**Comment by hgwood** [comment]
OK, found something that works a bit better IMO, since it's now possible to navigate to the stats page immediately.

**Pull request merged**